### PR TITLE
Add Debit and Debit Refund to transaction status codes table

### PR DESCRIPTION
**Pull request by cbobach**
Fixes https://github.com/clearhaus/issues-pci/issues/3509.

I've went through klovedal `factory.rb` and `visa-direct.rb` to conduct which status codes are reachable.

image

I'm a bit pulsed about the `Invalid currency` as I suppose that the currency is inherited from the original debit.

### CODE REVIEW

**Comment by mt-clearhaus** [comment]
Regarding invalid currency, we apply the same filter for captures, refunds and voids, even though they also inherit the currency. There is the corner case where support for a currency is dropped and a client wishes to capture an authorization with an unsupported currency. Similarly, a client could wish to refund a debit with an unsupported currency. So I actually think the correct solution is to indicate that invalid currency can be received for all transaction types.

File: website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
Comment (ct-clearhaus) [inline website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:17]: I was surprised that it's not possible for the debit refund, but I notice that one cannot supply a ToS for a debit refund — so it's different from the capture refund in this regard.

File: website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
Comment (ct-clearhaus) [inline website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:18]: It may not be possible at this time for debit refunds, but it will in the not too distant future, so let's just add it right away — it should be expected possible by the client.

```suggestion
|          | 40200 | Clearhaus rule violation           | ✓    | ✓       | ✓      | ✓    | ✓      | ✓      | ✓            |
```

File: website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
Comment (mt-clearhaus) [inline website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:11]:
```suggestion
|          | 40110 | Invalid card number                | ✓    |         |        |      | ✓      | ✓      | ✓            |
```

It sounds obscure, however, we cannot guarantee that we will not respond with invalid card number for a debit refund. We are talking to upstream, i.e. Visa, and an issuer, or Visa, might respond to us with ISO 8583 response code 14 meaning invalid card number.

It is quite unrealistic, though, as a debit refund must happen relatively quickly after a debit.

@ct-clearhaus WDYT?

File: website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
Comment (mt-clearhaus) [inline website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:15]: Same story as for 40110 / invalid card number. A debit and debit refund pair might cross end of month, potentially entailing card expiry. We could receive ISO 8583 response code 54 from upstream meaning card expired. Not very realistic, and perhaps Visa even allows debit refunds on expired cards.

File: website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
Comment (mt-clearhaus) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:11]: Issuers can potentially send us any ISO 8583 response code, and we would do the usual mapping expected of us.

Comment (mt-clearhaus) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:11]: Looking at the current 404xx status codes for debit refunds, it seems we lean towards putting a checkmark on all possible responses from upstream. So I'm leaning towards we do that consistently for all possible responses (i.e. also for invalid card number, invalid CSC, card expired, merchant blocked by cardholder).

However, it does feel weird to mention invalid CSC as a possibility for a debit refund.

Comment (mt-clearhaus) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:11]: I've striked "merchant blocked by cardholder" above as I see that's not a possible case for debits and debit refunds.

File: website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
Comment (mt-clearhaus) [inline website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:16]:
```suggestion
|          | 40140 | Invalid currency                   | ✓    | ✓       | ✓      | ✓    | ✓      | ✓      | ✓            |
```

See https://github.com/clearhaus/gateway-api-docs/pull/172#issuecomment-1527190607.

File: website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
Comment (ct-clearhaus) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:11]: The debit refund is a "live" transaction, so more response codes are actually possible, yes 👍

I didn't think about this 🙈

File: website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
Comment (mt-clearhaus) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:15]:
```suggestion
|          | 40135 | Card expired                       | ✓    |         |        |      | ✓      | ✓      | ✓            |
```

File: website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
Comment (mt-clearhaus) [inline website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:13]:
```suggestion
|          | 40120 | Invalid CSC                        | ✓    |         |        |      |        | ✓      | ✓            |
```

This one hurts.

File: website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
Comment (mt-clearhaus) [inline website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:7]:
```suggestion
| Status   | Code  | Meaning                            | Auth | Capture | Refund | Void | Credit | Debit  | Debit Refund |
```

**Comment by mt-clearhaus** [comment]
The current table is wider than the other parts of the site:

Screenshot from 2023-04-28 11-17-30

Suggestion:

<details>
<summary>Click to view diff</summary>

```diff
--- a/website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
+++ b/website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
@@ -4,27 +4,29 @@ date: 2022-04-13T12:37:22+02:00
 anchor: "transaction-status-codes-table"
 weight: 195
 ---
-| Status   | code  | Meaning                            | Auth | Capture | Refund | Void | Credit | Debit  | Debit Refund |
-|----------|-------|------------------------------------|------|---------|--------|------|--------|--------|--------------|
-| Approved | 20000 | Approved                           | ✓    | ✓       | ✓      | ✓    | ✓      | ✓      | ✓            |
-| Declined | 40000 | General input error                | ✓    | ✓       | ✓      | ✓    | ✓      | ✓      | ✓            |
-|          | 40110 | Invalid card number                | ✓    |         |        |      | ✓      | ✓      |              |
-|          | 40111 | Unsupported card scheme            | ✓    |         |        |      | ✓      | ✓      |              |
-|          | 40120 | Invalid CSC                        | ✓    |         |        |      |        | ✓      |              |
-|          | 40130 | Invalid expire date                | ✓    |         |        |      | ✓      | ✓      |              |
-|          | 40135 | Card expired                       | ✓    |         |        |      | ✓      | ✓      |              |
-|          | 40140 | Invalid currency                   | ✓    |         |        |      | ✓      | ✓      | ✓            |
-|          | 40150 | Invalid text on statement          | ✓    | ✓       | ✓      |      | ✓      | ✓      |              |
-|          | 40200 | Clearhaus rule violation           | ✓    | ✓       | ✓      | ✓    | ✓      | ✓      |              |
-|          | 40300 | 3-D Secure problem                 | ✓    |         |        |      |        | ✓      |              |
-|          | 40310 | 3-D Secure authentication failure  | ✓    |         |        |      |        | ✓      |              |
-|          | 40400 | Backend problem                    | ✓    |         |        | ✓    | ✓      | ✓      | ✓            |
-|          | 40410 | Declined by issuer or card scheme  | ✓    |         |        | ✓    | ✓      | ✓      | ✓            |
-|          | 40411 | Card restricted                    | ✓    |         |        |      | ✓      | ✓      | ✓            |
-|          | 40412 | Card lost or stolen                | ✓    |         |        |      | ✓      | ✓      | ✓            |
-|          | 40413 | Insufficient funds                 | ✓    |         |        |      | ✓      | ✓      | ✓            |
-|          | 40414 | Suspected fraud                    | ✓    |         |        |      | ✓      | ✓      | ✓            |
-|          | 40415 | Amount limit exceeded              | ✓    |         |        |      | ✓      | ✓      | ✓            |
-|          | 40416 | Additional authentication required | ✓    |         |        |      | ✓      | ✓      | ✓            |
-|          | 40420 | Merchant blocked by cardholder     | ✓    |         |        |      | ✓      |        |              |
-|          | 50000 | Clearhaus error                    | ✓    | ✓       | ✓      | ✓    | ✓      | ✓      | ✓            |
+| Code    | Meaning                            | Auth | Capture | Refund | Void | Credit | Debit  | Debit Refund |
+|---------|------------------------------------|------|---------|--------|------|--------|--------|--------------|
+| `20000` | Approved                           | ✓    | ✓       | ✓      | ✓    | ✓      | ✓      | ✓            |
+| `40000` | General input error                | ✓    | ✓       | ✓      | ✓    | ✓      | ✓      | ✓            |
+| `40110` | Invalid card number                | ✓    |         |        |      | ✓      | ✓      |              |
+| `40111` | Unsupported card scheme            | ✓    |         |        |      | ✓      | ✓      |              |
+| `40120` | Invalid CSC                        | ✓    |         |        |      |        | ✓      |              |
+| `40130` | Invalid expire date                | ✓    |         |        |      | ✓      | ✓      |              |
+| `40135` | Card expired                       | ✓    |         |        |      | ✓      | ✓      |              |
+| `40140` | Invalid currency                   | ✓    |         |        |      | ✓      | ✓      | ✓            |
+| `40150` | Invalid text on statement          | ✓    | ✓       | ✓      |      | ✓      | ✓      |              |
+| `40200` | Clearhaus rule violation           | ✓    | ✓       | ✓      | ✓    | ✓      | ✓      |              |
+| `40300` | 3-D Secure problem                 | ✓    |         |        |      |        | ✓      |              |
+| `40310` | 3-D Secure authentication failure  | ✓    |         |        |      |        | ✓      |              |
+| `40400` | Backend problem                    | ✓    |         |        | ✓    | ✓      | ✓      | ✓            |
+| `40410` | Declined by issuer or card scheme  | ✓    |         |        | ✓    | ✓      | ✓      | ✓            |
+| `40411` | Card restricted                    | ✓    |         |        |      | ✓      | ✓      | ✓            |
+| `40412` | Card lost or stolen                | ✓    |         |        |      | ✓      | ✓      | ✓            |
+| `40413` | Insufficient funds                 | ✓    |         |        |      | ✓      | ✓      | ✓            |
+| `40414` | Suspected fraud                    | ✓    |         |        |      | ✓      | ✓      | ✓            |
+| `40415` | Amount limit exceeded              | ✓    |         |        |      | ✓      | ✓      | ✓            |
+| `40416` | Additional authentication required | ✓    |         |        |      | ✓      | ✓      | ✓            |
+| `40420` | Merchant blocked by cardholder     | ✓    |         |        |      | ✓      |        |              |
+| `50000` | Clearhaus error                    | ✓    | ✓       | ✓      | ✓    | ✓      | ✓      | ✓            |
+
+Status code `20000` is the only code where the transaction is approved. For the other status codes the transaction is declined.
```
</details>

With the diff applied:

Screenshot from 2023-04-28 11-14-45
[...]
Screenshot from 2023-04-28 11-15-01

**Comment by cbobach** [comment]
I'm all for the change, maybe you can push it @mt-clearhaus? Then @ct-clearhaus can approve the PR if he agrees on the change :smile:

File: website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
Comment (ct-clearhaus) [inline website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:13]: Debit refunds can receive "Invalid CSC"? There should be no CSC sent, right? 🤨

File: website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
Comment (ct-clearhaus) [inline website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:25]: This is nonsense ("there's not enough money on the card to move money to the card"), but I absolutely do not doubt that we'll receive the response from some issuers, so 👍

**Review by ct-clearhaus — APPROVED** [review]
Minor non-important question, but otherwise I fully agree 👍 🙇‍♂️

File: website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
Comment (ct-clearhaus) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:25]: I'll resolve without further ado.

File: website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
Comment (cbobach) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md]: You are correct, fixed in [clearhaus/gateway-api-docs@`f734b37` (#172)](https://github.com/clearhaus/gateway-api-docs/pull/172/commits/f734b37e5390e22b2a37a8a2be93af4b59c3f4da)

Comment (mt-clearhaus) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md]: I was about to disagree, however, I see we do not currently map any responses from upstream to 40120 for debits and debit refunds.

But! We should start to perform such mapping. Especially for debits, where it indeed can make sense. Please await an internal code PR before moving on.

Comment (mt-clearhaus) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md]: With the internal code PR now created, I'll argue that both debits and debit refunds should have a :heavy_check_mark: for invalid CSC.

Comment (cbobach) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md]: But how should a client get an `Invalid CSC` triggered on a refund? A client can only refund a debit that has gone through this check and we use the csc value stored with the card. As a client you only send the debit id, right. :thinking: 

I agree that we could just add it there also because Visa could answer this to us.

Comment (mt-clearhaus) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md]: You write it yourself :slightly_smiling_face: 

> because Visa could answer this to us.

It is the same story as 40413 / insufficient funds, for which @ct-clearhaus commented here: https://github.com/clearhaus/gateway-api-docs/pull/172#discussion_r1182112969

Comment (mt-clearhaus) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md]: I do agree it is annoying we have to have all these checkmarks because issuers can potentially respond with all response codes. It almost ruins the value of the table.

Comment (mt-clearhaus) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:13]: @ct-clearhaus Do you have any final input to this?

Comment (cbobach) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:13]: I've reverted the before mentioned Debit refund Invalid CSC marking in [`b45b9b0` (#172)](https://github.com/clearhaus/gateway-api-docs/pull/172/commits/b45b9b03a8c3118158a9b5d33c59ecf638b293c8)

Comment (ct-clearhaus) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:13]: You are simply awesome! 🤗

File: website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md
Comment (mt-clearhaus) [inline website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:29]:
```suggestion
| 40420 | Merchant blocked by cardholder     | ✓    |         |        |      | ✓      | ✓      | ✓            |
```

Because we can potentially receive `R0` and `R2` from upstream for debits and and debit refunds.

Comment (mt-clearhaus) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:29]: Sadly.

Comment (mt-clearhaus) [reply website/content/gateway/api_reference/transaction_status/transaction_status_codes_table.md:29]: I hope this is my last suggestion :crossed_fingers:

**Review by mt-clearhaus — APPROVED** [review]

**Comment by mt-clearhaus** [comment]
I believe we are ready to merge this :rocket: :1st_place_medal:

**Pull request merged**
